SEM-146 Make printers country code lockable in production

onnodb opened the pull request:
_This PR is part of Ultimaker/opinicus#2003, and implements changes to the kernel boot script._

The current [printer provisioning process](https://ultimaker.atlassian.net/l/cp/oRBu1zNW) is documented quite extensively on our wiki. It basically boils down the following:

Early on in the boot process, the kernel init script (`initramfs/init.sh` in the repo) performs a bunch of tasks. One of these checks if a USB drive has been inserted. If that USB drive contains a single ext4 partition, it can be mounted. If the script then finds an `article_number` file in the drive  root, it programs the BOM number value in the mainboard EEPROM with the contents of the file.

We extend this by allowing an additional file to be present in the USB drive's root, `country_code_lock`. The contents of this file will be written to addres `0x0118` in the mainboard EEPROM, from where it will be picked up by the opinicus firmware in Ultimaker/opinicus#2003.

To provision a printer so that it's locked to Australia, you'd take a USB stick with a file `country_code_lock` that contains a single line of text:
```
0x41 0x55
```
To wipe the Australia lock from a printer, you'd put a file `country_code_lock` with a single line:
```
0xFF 0xFF
```
To provision a printer in a way that simply leaves the country lock as-is, just don't put this file on the USB drive.

(Do keep in mind that not all country codes are valid at this point; [see here](https://github.com/Ultimaker/opinicus/blob/d47f1813c67e589ef1cc4717cc4bfe56940e5f84/griffin/system/eeprom.py#L40-L50))

onnodb committed:
SEM-146 Write country lock during EEPROM provisioning

Extend the printer provisioning logic in the `init.sh` boot script, so
that a "locked country code" can be set in the mainboard EEPROM.

The existing provisioning code looks for an `article_number` file on a
USB drive that's inserted during boot (ext4, single partition). The
contents of that file, a single line of text in hex notation describing
the BOM number, is then written to the mainboard EEPROM using
`i2ctransfer`.

We duplicate this code, so that an additional file `country_code_lock`
can be used to also write a "locked country code" to the EEPROM, at
address 0x0118. This file is optional, so default provisioning keeps
whatever data is here, intact. The file works the same way: it should
contain a single line of text, with a hex representation of the locked
country code.

onnodb committed:
SEM-146 Document provisioning logic in the init.sh boot script

Add a comment to the `init.sh` boot script, explaining the logic behind
the EEPROM provisioning.

Jipolie01 commented on initramfs/init.sh:
```diff
@@ -218,15 +218,44 @@ isBootingRestoreImage()
+    # If either of these files is missing, no data is written to that part of the EEPROM.
+    #
+    # Relevant EEPROM layout here:
+    #   0x0100 - 0x0104     BOM number / article number
+    #   0x0118 - 0x0120     Locked country code
```
Small nitpick, but between 108 and 118 the guid of the machine can be found. I think it would be nice to add since that will remove any confusion that there is 'space' between these two addresses. But up to you if you want to do that

onnodb replied:
I don't intend to duplicate the full EEPROM layout here, as that's already documented in Opinicus, and I want to avoid having multiple sources of truth. I do agree that it's good to avoid confusion about the stuff between 0x0104 and 0x0118, so I've added an ellipsis in between, plus a reference to opinicus as the source of truth.

Jipolie01 commented on initramfs/init.sh:
```diff
@@ -241,19 +270,32 @@ check_and_set_article_number()
+            # shellcheck disable=SC2086
+            if ! i2ctransfer -y 3 w4@0x57 0x01 0x18 ${country_code}; then
+                umount "${dev}"
+                echo "Failed to write country code lock to EEPROM, skipping."
```
Do we need a way to tell ZB that this has failed or will they check the machine before it goes out?

onnodb replied:
They will check the result by looking at the updated "Network" menu in Ultimaker/okuda#871: that will show a successfully locked country.

At this point of system boot, I don't think we have a good way of communicating stuff back to the user...

Jipolie01 reviewed (approved):
Cool stuff, a couple of small things that don't block this from merging. Up to you to see if you want to merge it

alexborro commented on initramfs/init.sh:
```diff
@@ -241,19 +270,32 @@ check_and_set_article_number()
+            echo "Trying to write article nr: '${article_number}'."
+            # shellcheck disable=SC2086
+            if ! i2ctransfer -y 3 w6@0x57 0x01 0x00 ${article_number}; then
+                umount "${dev}"
```
You cannot unmount the usb stick at this point, otherwise you cannot read the country code below.
I would do the first 2 checks and, if file found, do your thing, otherwise output a message saying file not found and, in the end, umount the usb stick.

I added the branch `personal/alex/SEM-146-country_code_lock` with this change, but it looks like this:

```
        echo "Attempting to mount '${dev}'."
        if ! mount -t f2fs,ext4,vfat,auto -o exec,noatime "${dev}" "${ARTICLENR_USB_MOUNT}"; then
            return 0
        fi

        if [ -r "${article_number_file}" ]; then
            article_number="$(cat "${article_number_file}")"
            echo "Trying to write article nr: '${article_number}'."
            # shellcheck disable=SC2086
            if ! i2ctransfer -y 3 w6@0x57 0x01 0x00 ${article_number}; then
                echo "Failed to write article number to EEPROM, skipping."
            fi
        else
            echo "No article number file found on '${dev}', skipping."
        fi

        if [ -r "${country_code_lock_file}" ]; then
            country_code="$(cat "${country_code_lock_file}")"
            echo "Trying to write country code lock: '${country_code}'."
            # shellcheck disable=SC2086
            if ! i2ctransfer -y 3 w4@0x57 0x01 0x18 ${country_code}; then
                echo "Failed to write country code lock to EEPROM, skipping."
            fi
        else
            echo "No country code lock file found on '${dev}', skipping."
        fi

        umount "${dev}"
        retries=0
```

onnodb replied:
Thanks again @alexborro . Incorporated in e160b4c24239d97d5adeefd03caac937005d624e

alexborro commented on initramfs/init.sh:
```diff
@@ -218,15 +218,44 @@ isBootingRestoreImage()
+
     dev="/dev/sda1"
     article_number_file="${ARTICLENR_USB_MOUNT}/article_number"
+    country_code_lock_file="${ARTICLENR_USB_MOUNT}/country_code_lock"
```
I would rename the `ARTICLENR_USB_MOUNT` to something like `PROVISIONING_USB_MOUNT` since now it have more meaning than only article number.

onnodb replied:
Yes, good point!

onnodb committed:
SEM-146 Boyscout: improve init.sh script

Improve the code in the `check_and_set_eeprom_data()` part of the
`init.sh` boot script. Based on feedback from @alexborro:

- No longer let the while loop span the entire code within the function.
If you analyze the purpose and flow of the while loop, it turns out it's
only ever looping to wait for the USB drive to show up under the `/dev/`
directory. All other code paths lead to either a `return` or the end of
the while loop. That makes things unnecessarily confusing.

- If an `i2ctransfer` call fails, don't exit the function completely,
but give the *other* provisioning file another chance. Do make sure that
all failures and skips are properly logged.

Co-authored-by: Alex Borro <alexborro@gmail.com>

atrebelo commented:
### SUPER IMPORTANT INFORMATION
⚠️ As this change is for the S7, this change needs to be cherry-picked to `master/imx6dl-03E` - as this is the `um-kernel` we are using for the S7.

Ignore ☝🏻 ; I got my wires crossed when releasing `um-kernel` from the `master/imx6dl` branch, we need to use the `8.1.3-imx6dl-03E` for the version.
